src/nsec3_fmt_plug.c

```diff
@@ -0,0 +1,396 @@
+/*
+ * DNSSEC NSEC3 hash cracker. Developed as part of the nsec3map DNS zone
+ * enumerator project (https://github.com/anonion0/nsec3map).
+ *
+ * This software is Copyright (c) 2016 Ralf Sager <nsec3map at 3fnc.org>,
+ * and it is hereby released to the general public under the following terms:
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Some of this code was inspired by sha1_gen_fmt_plug.c by Solar Designer
+ */
+
+#if FMT_EXTERNS_H
+extern struct fmt_main fmt_nsec3;
+#elif FMT_REGISTERS_H
+john_register_one(&fmt_nsec3);
+#else
+
+#include <ctype.h>
+#include <string.h>
+#include <stdint.h>
+#include "sha.h"
+
+#include "arch.h"
+#include "params.h"
+#include "common.h"
+#include "formats.h"
+#include "memdbg.h"
+
+#define FORMAT_LABEL                    "nsec3"
+#define FORMAT_NAME                     "DNSSEC NSEC3"
+#define ALGORITHM_NAME                  "32/" ARCH_BITS_STR
+#define BENCHMARK_COMMENT               ""
+#define BENCHMARK_LENGTH                0
+#define PLAINTEXT_LENGTH                125
+#define MIN_KEYS_PER_CRYPT              1
+#define MAX_KEYS_PER_CRYPT              1
+#define BINARY_SIZE                     20
+#define BINARY_ALIGN                    sizeof(ARCH_WORD_32)
+#define N3_MAX_SALT_SIZE                255
+#define N3_MAX_ZONE_SIZE                255
+#define HASH_LENGTH                     20
+#define SALT_SIZE                       sizeof(struct salt_t)
+#define SALT_ALIGN                      sizeof(ARCH_WORD_32)
+#define FORMAT_TAG                      "$NSEC3$"
+#define FORMAT_TAG_LENGTH               (sizeof(FORMAT_TAG) - 1)
+
+struct salt_t {
+	uint16_t iterations;
+	size_t salt_length;
+	size_t zone_length;
+	unsigned char salt[N3_MAX_SALT_SIZE];
+	unsigned char zone_wf[N3_MAX_ZONE_SIZE];
+};
```

Comment:
To avoid padding I think it should be like this (or would any compiler silently move them around?):
```c
struct salt_t {
	size_t salt_length;
	size_t zone_length;
	uint16_t iterations;
	unsigned char salt[N3_MAX_SALT_SIZE];
	unsigned char zone_wf[N3_MAX_ZONE_SIZE];
};
```